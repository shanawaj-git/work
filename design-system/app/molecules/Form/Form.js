import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';

export const Form = ({ form, InputComponent, formState }) => {
  const [inputs, setInputs] = useState({});

  useEffect(() => {
    const formObject = Object.values(form).reduce((acc, formElement) => {
      const { name, value } = formElement;
      acc[name] = value;
      return acc;
    }, {});

    setInputs(formObject);
  }, [form]);

  useEffect(() => {
    formState(inputs);
  }, [inputs]);
  return (
    <>
      {Object.values(form).map(formElement => (
        <InputComponent
          key={formElement.name}
          {...formElement}
          onChange={e => {
            setInputs({ ...inputs, [formElement.name]: e.target.value });
          }}
        />
      ))}
    </>
  );
};

// Props Type
Form.propTypes = {
  form: PropTypes.shape({
    [PropTypes.string]: PropTypes.shape({
      name: PropTypes.string,
      value: PropTypes.string,
    }),
  }),
  InputComponent: PropTypes.any,
  formState: PropTypes.func,
};
