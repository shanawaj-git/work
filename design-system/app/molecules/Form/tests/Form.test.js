import React from 'react';
import renderer, { act } from 'react-test-renderer';
import { unmountComponentAtNode } from 'react-dom';
import { fireEvent, render } from 'react-testing-library';

import Form from '../index';

let container = null;

describe('Form.js', () => {
  const props = {
    form: {
      firstName: {
        name: 'firstName',
        value: 'Ahmed',
      },
      city: {
        name: 'city',
        value: 'Alexandria',
      },
    },
    InputComponent: () => <input {...props} />,
    formState: jest.fn(),
  };

  beforeEach(() => {
    // setup a DOM element as a render target
    container = document.createElement('div');
    document.body.appendChild(container);
  });

  afterEach(() => {
    // cleanup on exiting
    unmountComponentAtNode(container);
    container.remove();
    container = null;
  });

  describe('index', () => {
    it('Should render inputs based on the form prop', () => {
      const tree = renderer.create(<Form {...props} />);

      renderer.create(<Form {...props} />);
      expect(tree).toBeTruthy();
      expect(tree).toMatchSnapshot();
    });

    it('Should call onChange if input field changed', async () => {
      props.onChange = jest.fn();
      act(() => {
        render(<Form {...props} />);
      });
      const input = document.querySelector('input');
      const newValue = 'update field';
      fireEvent.change(input, { target: { value: `${newValue}` } });
      expect(input.value).toEqual(`${newValue}`);
      expect(props.onChange).toHaveBeenCalledTimes(1);
    });
  });
});
