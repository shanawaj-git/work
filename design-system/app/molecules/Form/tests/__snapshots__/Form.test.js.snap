// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form.js index Should render inputs based on the form prop 1`] = `
Array [
  <input
    InputComponent={[Function]}
    form={
      Object {
        "city": Object {
          "name": "city",
          "value": "Alexandria",
        },
        "firstName": Object {
          "name": "firstName",
          "value": "Ahmed",
        },
      }
    }
    formState={
      [MockFunction] {
        "calls": Array [
          Array [
            Object {},
          ],
          Array [
            Object {
              "city": "Alexandria",
              "firstName": "Ahmed",
            },
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      }
    }
  />,
  <input
    InputComponent={[Function]}
    form={
      Object {
        "city": Object {
          "name": "city",
          "value": "Alexandria",
        },
        "firstName": Object {
          "name": "firstName",
          "value": "Ahmed",
        },
      }
    }
    formState={
      [MockFunction] {
        "calls": Array [
          Array [
            Object {},
          ],
          Array [
            Object {
              "city": "Alexandria",
              "firstName": "Ahmed",
            },
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      }
    }
  />,
]
`;
